compiler/hash-parser/src/lexer.rs

```diff
@@ -0,0 +1,618 @@
+                // >---------------------------------<
+                '(' => break TokenKind::OpenParen,
+                ')' => break TokenKind::CloseParen,
+                '{' => break TokenKind::OpenBrace,
+                '}' => break TokenKind::CloseBrace,
+                '[' => break TokenKind::OpenBracket,
+                ']' => break TokenKind::CloseBracket,
+                // >---------------------------------<
+                '~' => break TokenKind::Tilde,
+                '=' => break TokenKind::Eq,
+                '!' => break TokenKind::Exclamation,
+                '<' => break TokenKind::Lt,
+                '>' => break TokenKind::Gt,
+                '-' => break TokenKind::Minus,
+                '&' => break TokenKind::And,
+                '|' => break TokenKind::Pipe,
+                '+' => break TokenKind::Plus,
+                '*' => break TokenKind::Star,
+                '^' => break TokenKind::Caret,
+                '%' => break TokenKind::Percent,
+
+                // @@Improvement: This can be a potentially made out as a compound token... the same could be done with other
+                //                tokens that are likely to be compound, it would also avoid doing the work later on...
+                // ':' => match self.peek() {
+                //     ':' => {
+                //         self.next();
+                //         break TokenKind::NameAccess
+                //     }
+                //     _ => break TokenKind::Colon
+                // },
+                ':' => break TokenKind::Colon,
+
+                // Identifier (this should be checked after other variant that can
+                // start as identifier).
+                c if is_id_start(c) => break self.ident(),
+
+                // Numeric literal.
+                '0'..='9' => {
+                    break self
+                        .number()
+                        .unwrap_or_else(|e| panic!("error: {:#?}", e.message))
+                }
+                // character literal.
+                '\'' => {
+                    break self
+                        .char()
+                        .unwrap_or_else(|e| panic!("error: {:#?}", e.message)); // @@ErrorReporting: this is where we hook into error reporting to print the result
+                                                                                // and display it to the user
+                }
+
+                // String literal.
+                '"' => break self.string().unwrap_or_else(|e| panic!("error: {:?}", e)),
+
+                _ => break TokenKind::Unexpected,
+            }
+        };
+
+        let location = Location::span(offset, self.len_consumed());
+        Some(Token::new(token_kind, location))
+    }
+
+    /// Consume an identifier, at this stage keywords are also considered to be identfiers
+    pub(crate) fn ident(&self) -> TokenKind {
```

Comment:
Maybe it should be documented that this expects the first character to have already been consumed?